File: runtimes/swiftui/Sources/StyleConverterRuntime/Renderer/InlineRunPlan.swift
import Foundation

/// InlineRunPlan — the wave-32 `meta.runs` resolver (lane R), the SwiftUI
/// twin of web's `renderer/InlineRuns.ts` and Compose's `InlineRunPlan.kt`.
///
/// WHAT THE WIRE SAYS (schema/spec/03-children.md §4.1): `meta.runs` is the
/// component's inline content in DOCUMENT order — entries that are each
/// exactly one of `{text}` (an anonymous inline run) or `{child}` (the
/// position of one child box). It exists because `text` is ONE string and
/// the child list has ONE order, so the wire could previously say
/// `run + children` or `children + run` but never `run / child / run` —
/// exactly what `the quick <u>brown</u> fox` needs when the `<u>` survives.
///
/// WHAT THIS FILE DOES: fold the wire list into the shape this renderer can
/// consume WITHOUT restructuring its child `ForEach` — a per-child list of
/// the text runs that precede it, plus the runs that trail the last child.
/// That fold is only valid when the plan's child references appear in the
/// SAME order as the rendered children, so `resolve` proves it and refuses
/// otherwise (see `strictlyIncreasing` below). Refusing returns nil, which
/// puts the container back on the exact pre-wave-32 path.
///
/// HONEST SCOPE ON THIS PLATFORM. The block container is a `VStack`: it has
/// no inline line box, so an anonymous text run is a STACKED label, not a
/// fragment sharing a line with the child beside it. `meta.runs` therefore
/// buys SwiftUI the correct ORDER, not correct inline layout — the same
/// approximation the leading-`text` label always was, now at the right
/// position instead of always first. The CSS2 static-position family is
/// decided purely by order, so that is a real fidelity gain; the full
/// inline-formatting-context model stays out of scope.
enum InlineRunPlan {

    /// The folded plan.
    struct Plan {
        /// Text runs that precede child `i`, keyed by that child's index.
        let before: [Int: [String]]
        /// Text runs after the last referenced child.
        let trailing: [String]
    }

    /// Resolve `runs` against the rendered `children` (already `order`-sorted,
    /// in the exact sequence the caller's `ForEach` walks).
    ///
    /// - Returns: the folded plan, or nil meaning "no usable runs — keep the
    ///   default leading-text-then-children path". nil rather than an empty
    ///   plan is what keeps every pre-wave-32 document's view tree identical:
    ///   absent, malformed, fully-dangling and out-of-order lists all land
    ///   back on the code path they used before.
    static func resolve(_ runs: [IRRun]?, children: [IRComponent]) -> Plan? {
        // Omit-when-absent is the overwhelming common case.
        guard let runs, !runs.isEmpty, !children.isEmpty else { return nil }

        // Two lookup keys, because the reference is the child's AUTHORING KEY
        // (spec 03 §4.1): on the converter-emitted wire that key is the
        // child's `name`; in the extractor-direct pipeline it is also its
        // `id`. `name` is consulted first so a document that happens to reuse
        // a name as some other component's id cannot mis-resolve.
        var byName: [String: Int] = [:]
        var byId: [String: Int] = [:]
        for (index, child) in children.enumerated() {
            // First occurrence wins on a duplicate — the composer is lenient
            // the same way (duplicate ids are a convert-time error upstream).
            if !child.name.isEmpty, byName[child.name] == nil { byName[child.name] = index }
            if !child.id.isEmpty, byId[child.id] == nil { byId[child.id] = index }
        }

        var before: [Int: [String]] = [:]
        var trailing: [String] = []
        var pending: [String] = []
        var claimedOrder: [Int] = []
        var sawChild = false

        for run in runs {
            if let text = run.text {
                // An EMPTY run contributes no glyphs and no box. A
                // WHITESPACE-ONLY run is not empty — it is the inter-run word
                // space (rule 6) and must survive.
                if !text.isEmpty { pending.append(text) }
                continue
            }
            guard let key = run.child else { continue } // decoder rejected this
            guard let index = byName[key] ?? byId[key] else {
                // Rule 5 — dangling key: skip. The named box is not in this
                // composition, so there is nothing to place; every other
                // entry keeps its position.
                continue
            }
            // A child named twice would render twice; honour the FIRST.
            if claimedOrder.contains(index) { continue }
            claimedOrder.append(index)
            sawChild = true
            if !pending.isEmpty { before[index] = pending; pending = [] }
        }
        // Whatever is still pending sits after the last referenced child.
        trailing = pending

        // Nothing to place, or nothing but text: fall back. A list with no
        // child reference says nothing the plain `text` channel does not.
        guard sawChild, !(before.isEmpty && trailing.isEmpty) else { return nil }

        // THE PROOF this fold needs: the caller walks children in index order,
        // so a plan whose child references are NOT in that order cannot be
        // expressed as "runs before child i". Rather than silently paint the
        // wrong sequence, refuse — the container keeps its pre-wave-32
        // behaviour and the fixture still carries `_runs` for the platforms
        // that can express it. (The extractor emits document order and the
        // children array is document order, so this holds for the whole
        // committed corpus; the guard is here because `FlexboxApplier.sorted`
        // may permute children by CSS `order`.)
        guard strictlyIncreasing(claimedOrder) else { return nil }
        return Plan(before: before, trailing: trailing)
    }

    /// Is every element greater than the one before it?
    private static func strictlyIncreasing(_ xs: [Int]) -> Bool {
        for i in 1..<max(xs.count, 1) where xs[i] <= xs[i - 1] { return false }
        return true
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/core/renderer/InlineRunPlan.kt
package com.styleconverter.runtime.core.renderer

import com.styleconverter.runtime.core.ir.IRComponent
import com.styleconverter.runtime.core.ir.IRRun

/**
 * InlineRunPlan — the wave-32 `meta.runs` resolver (lane R), the Compose
 * twin of web's `renderer/InlineRuns.ts` and iOS's `InlineRunPlan.swift`.
 *
 * WHAT THE WIRE SAYS (schema/spec/03-children.md §4.1): `meta.runs` is the
 * component's inline content in DOCUMENT order — entries that are each
 * exactly one of `{text}` (an anonymous inline run) or `{child}` (the
 * position of one child box). It exists because `_text` is ONE string and
 * the child list has ONE order, so the wire could previously say
 * `run + children` or `children + run` but never `run / child / run` —
 * exactly what `the quick <u>brown</u> fox` needs when the `<u>` survives.
 *
 * WHAT THIS FILE DOES: turn the wire list into indices into the ALREADY
 * COMPOSED `component.children`, applying the three rules a renderer owns:
 *   rule 2  a referenced child renders at its run slot and NOT again in the
 *           sibling walk (hence [unreferenced], not "all children");
 *   rule 4  a child the list does not name still renders, after the runs,
 *           in sibling order — a partial list can never lose a box;
 *   rule 5  a dangling `child` key is a warn-and-skip, never a throw.
 *
 * HONEST SCOPE ON THIS PLATFORM. Compose's block container is a `Column`:
 * it has no inline line box, so an anonymous text run is a STACKED text
 * box, not a fragment sharing a line with the child beside it. `meta.runs`
 * therefore buys Compose the correct ORDER, not correct inline layout —
 * the same approximation the leading-`_text` box always was, now at the
 * right position instead of always first. That is a real fidelity gain
 * (the CSS2 static-position family is decided purely by order) and it is
 * NOT the full inline-formatting-context model.
 *
 * Wave 44 (lane U1) AMENDS that scope for ONE member family: when every
 * resolved member is a plain text run or a policy-only/empty inline leaf,
 * `typography/inline/InlineRunFold` collapses the plan into a single
 * paragraph rendered through the ordinary one-Text pipeline — real
 * inline flow (shared line boxes, one greedy break pass, one baseline)
 * for exactly the shapes where one paragraph is provably equivalent to
 * the inline formatting context. Everything else (atoms, floats, abspos,
 * styled/nested inline boxes) still takes the stacked fallback above,
 * with the refusal logged at the renderer seam.
 */
object InlineRunPlan {

    /** One resolved entry: a text run, or an index into `component.children`. */
    sealed interface Entry {
        data class Text(val text: String) : Entry
        data class Child(val index: Int) : Entry
    }

    /** The render plan: ordered content plus the children it left over. */
    data class Plan(val entries: List<Entry>, val unreferenced: List<Int>)

    /**
     * Resolve [runs] against [children].
     *
     * @return the plan, or null meaning "no usable runs — take the default
     *   leading-text-then-children path". Returning null rather than an
     *   empty plan is what keeps every pre-wave-32 document's composition
     *   byte-identical: absent, malformed and fully-dangling lists all land
     *   back on the exact code path they used before.
     */
    fun resolve(runs: List<IRRun>?, children: List<IRComponent>?): Plan? {
        // Omit-when-absent is the overwhelming common case — bail before
        // allocating anything.
        if (runs.isNullOrEmpty() || children.isNullOrEmpty()) return null

        // Two lookup keys, because the reference is the child's AUTHORING
        // KEY (spec 03 §4.1): on the converter-emitted wire that key is the
        // child's `name`; in the extractor-direct pipeline it is also its
        // `id`. `name` is consulted first so a document that happens to
        // reuse a name as some other component's id cannot mis-resolve.
        val byName = HashMap<String, Int>()
        val byId = HashMap<String, Int>()
        children.forEachIndexed { index, child ->
            // First occurrence wins on a duplicate — the composer is lenient
            // the same way (duplicate ids are a convert-time error upstream).
            if (child.name.isNotEmpty()) byName.putIfAbsent(child.name, index)
            if (child.id.isNotEmpty()) byId.putIfAbsent(child.id, index)
        }

        val entries = ArrayList<Entry>(runs.size)
        val claimed = HashSet<Int>()
        for (run in runs) {
            val text = run.text
            val key = run.child
            if (text != null) {
                // An EMPTY run contributes no glyphs and no box; dropping it
                // keeps the composition free of zero-content Text nodes. A
                // WHITESPACE-ONLY run is not empty — it is the inter-run word
                // space (rule 6) and must survive.
                if (text.isNotEmpty()) entries.add(Entry.Text(text))
                continue
            }
            if (key == null) continue // decoder already rejected this shape
            val index = byName[key] ?: byId[key]
            if (index == null) {
                // Rule 5 — dangling key: skip loudly. The named box is not in
                // this composition, so there is nothing to place; every other
                // entry keeps its position.
                // runCatching guards android.util.Log for JVM callers (the
                // unit suite exercises this pure resolver directly) — the
                // same guard logCollapseFallbackOnce uses.
                runCatching {
                    android.util.Log.w(
                        "InlineRunPlan",
                        "meta.runs names child '$key', which is not among the composed children — entry skipped"
                    )
                }
                continue
            }
            if (!claimed.add(index)) {
                // A child named twice would render twice; the wire has no
                // meaning for that, so honour the FIRST position and say so.
                runCatching {
                    android.util.Log.w(
                        "InlineRunPlan",
                        "meta.runs names child '$key' more than once — later reference skipped"
                    )
                }
                continue
            }
            entries.add(Entry.Child(index))
        }

        // Nothing survived: fall back to the default path rather than render
        // an empty container. The honest failure mode — the wire asked for an
        // order we could not reconstruct, so paint the pre-wave-32
        // approximation instead of painting nothing.
        if (entries.isEmpty()) return null

        // Rule 4 — leftovers, in sibling order, after the runs.
        val unreferenced = children.indices.filter { it !in claimed }
        return Plan(entries, unreferenced)
    }
}
